src/main/java/edu/kit/compiler/semantic/SymbolTable.java

```diff
@@ -0,0 +1,156 @@
+         * @param name The name of the symbol
+         * @return If there exists a symbol in the string table return the symbol otherwise create a new symbol
+         */
+        public Symbol insert(int name) {
+            if(table.get(name) == null) {
+                Symbol s = new Symbol(name);
+                table.put(name, s);
+                return s;
+            }
+            else {
+                return table.get(name);
+            }
+        }
+
+        public void remove(int name) {
+            table.remove(name);
+        }
+    }
+
+    /**
+     * Enter scope should be called when symbols from outer definitions can coexist with new definitions.
+     * <b>Note that enter scope has to be called once before interacting with the symbol table.</b>
+     */
+    public void enterScope() {
+        currentScope = new Scope(currentScope, changes.size());
+    }
+    public void leaveScope() {
+        // revert all previous changes
+        while (changes.size() > currentScope.oldSize) {
+            Change c = changes.pop();
+            // check if we leave our defining scope then remove symbol entirely from table
+            if (c.previousScope == null) {
+                symbolStringTable.remove(c.sym.name);
+                return;
+            }
+            c.sym.currentDefinition = c.previousDefinition;
+            c.sym.currentScope = c.previousScope;
+        }
+        currentScope = currentScope.parent;
+    }
+    public Symbol insert(Definition definition) {
+        int name = definition.getName();
+        Symbol s = symbolStringTable.find(name);
+        if(isDefinedInCurrentScope(s)) {
+            // this should never happen, make sure to check in upper level to get line number and column
+            throw new SemanticException("symbol is already defined in scope!", new CompilerException.SourceLocation(
+                    definition.getLine(),
+                    definition.getColumn()));
+        }
+        if(s == null) {
+            s = symbolStringTable.insert(name);
+        }
+
+        s.currentDefinition = null;
+        s.currentScope = null;
+
+        // last change resets the symbol
+        changes.push(new Change(s, null, null));
+        s.currentScope = currentScope;
+        s.currentDefinition = definition;
+
+        return s;
+    }
+    public Definition lookup(Symbol symbol) {
+        return symbol.currentDefinition;
+    }
+
+    /**
+     * Return a definition for this symbol if there exists one otherwise return null
+     * @param symbol The symbol to be checked
+     * @return Definition if there exists a symbol otherwise null
+     */
+    public Definition lookup(int symbol) {
+        Symbol s = symbolStringTable.find(symbol);
+        if(symbolStringTable.find(symbol) == null) {
+            return null;
+        }
+        else {
+            return s.currentDefinition;
+        }
+    }
+
+    public boolean isDefinedInCurrentScope(Symbol symbol) {
```

Comment:
Shouldn't this take an `int` as argument? What I want to do is check if my local variable is defined in the current scope, having its name. If not, where would I get the symbol from, in order to call this method?

Reply:
Changed it to take an integer that represents the symbols name